README.md
ananoosxbmc
===========

A XBMC plugin for Ananoos.
 
The plugin itself WILL NOT WORK as the new version of Ananoos is still pending release.

DO NOT TRY THIS PLUGIN and complain about it. 

Thanks

Axel

changelog.txt
[B]Version 0.0.1[/B]

- First version

default.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
   Ananoos plugin for XBMC
   Copyright (C) 2014 Axel de Vignon
   License: Apache 2.0 (See license.txt)
'''


import sys
import xbmc
import xbmcvfs
import xbmcgui
import xbmcplugin
import xbmcaddon
import time
import urllib
import re
import urlparse


# XBMC hooks
xbmc_version= xbmc.getInfoLabel('System.BuildVersion')
xbmc_date	= xbmc.getInfoLabel('System.BuildDate')
addon		= xbmcaddon.Addon(id='plugin.video.ananoos')
language	= addon.getLocalizedString
addon_path	= addon.getAddonInfo('path');
debug		= True
base_url	= sys.argv[0]
addon_handle= int(sys.argv[1])
args		= urlparse.parse_qs(sys.argv[2][1:])
mode		= args.get('mode', None)


# Plugin constants
version		= "0.1"
plugin		= "Ananoos/" + version
author		= "Axel de Vignon"
url			= "www.ananoos.com"
API_URL		= 'http://www.ananoos.com/api'


# Importing our librairies
import resources.lib.Auth as Auth
Auth = Auth.Auth()
import resources.lib.Utils as Utils	
import resources.lib.Api as Api
Api = Api.Api()



#Auth.Logout()
#Auth.Login()

def browse():
	xbmcplugin.setContent(addon_handle, 'movies')
	xbmcplugin.addSortMethod(addon_handle, 1)

	# Get current folder UID
	parent = args.get('parent', None)
	if parent is None:
		parent = '0'
	else:
		parent = str(parent[0])


	# Building API URL
	Utils.log('Getting items from API')

	try:
		content = Api.Browse(parent=parent)
		if content and content['list']:
			# Adding items to the listing
			for item in content['list']:
				# Item is a folder
				if item['type'] == 'dir':
					# Building URL
					url = base_url + '?' + urllib.urlencode({
						'mode' : 'browse', 
						'parent': item['id']
					})					

					Utils.log('Adding folder to list: '+url)
					li = xbmcgui.ListItem(item['name'], iconImage='DefaultFolder.png')
					xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li, isFolder=True)
				else:
					# Item is a regular file
					url = base_url + '?' + urllib.urlencode({
						'mode' : 'stream',
						'file': item['id']
					})				

					li = xbmcgui.ListItem(item['name'], iconImage='DefaultVideo.png')
					Utils.log('Adding file to list: '+url)
					xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li)
		xbmcplugin.endOfDirectory(addon_handle)
	except:
		Utils.log('Error when listing items')
		xbmc.executebuiltin('Notification('+language(30021)+', '+language(30022)+',3000,' + addon_path + '/resources/images/error.png)')
	

def stream():
	item_id = args.get('file', '0')
	Utils.log('Getting stream URL for file id: '+item_id[0])

	content = Api.getStreamUrl(item_id[0])
	if content and content['result'] == True:
		player = xbmc.Player()
		Utils.log('Stream URL is: '+content['url'])

		content['url'] = content['url'] + '|User-Agent='+urllib.quote(Utils.getUserAgent())+'&'

		# Sending cookies with the query
		cookies = Api.getCookies()
		if len(cookies) > 0:
			content['url'] = content['url'] + 'Cookie='
			for i in cookies:
				content['url'] = content['url'] + urllib.quote(i + '=' + cookies[i]+';')

		# Starting the playback
		player.play(content['url'])

		# If video doesn't work, raising an error
		time.sleep(3)
		if not player.isPlayingVideo():
			xbmc.executebuiltin('Notification('+language(30019)+', '+language(30020)+',3000,' + addon_path + '/resources/images/error.png)')
		
	else:
		xbmc.executebuiltin('Notification('+language(30019)+', '+language(30020)+',3000,' + addon_path + '/resources/images/error.png)')


def logout():
	Auth.logout()


''' Now routing '''
if mode is None or mode[0] == 'browse':
	browse()
elif mode[0] == 'stream':
	stream()
elif mode[0] == 'logout':
	logout()

resources/lib/Auth.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
   Ananoos plugin for XBMC
   Copyright (C) 2014 Axel de Vignon
   License: Apache 2.0 (See license.txt)
'''


import sys
import xbmc, xbmcgui
import resources.lib.Utils as Utils


default = sys.modules['__main__']

# Handling Auth
class Auth():

	def __init__(self):
		self.addon 			= default.addon
		self.username		= self.addon.getSetting('username') or ''
		self.password		= self.addon.getSetting('password') or ''
		self.window			= ''

	def Login(self, username = None, password = None):

		Utils.log('Checking login now')
		modal = 0

		# Checking whether arguments were sent
		if username is not None:
			self.username = username

		if password is not None:
			self.password = password

		# If username OR password are NOT available, opening modal box
		if self.username is None or self.password is None:
			Utils.log('Never logged in before, displaying modal')
			self.window = Utils.AuthForm(default.language(30006))
			self.window.doModal()
			return

		try:
			content = default.Api.Login(username=self.username, password=self.password)
			if not content or content == False:
					return False

			if content['result'] and content['result'] == 'success':
				#xbmc.executebuiltin('Notification('+default.language(30015)+', '+language(30016)+',3000,' + default.addon_path +'/resources/images/success.png)')
				Utils.log('Login is successful.')

				# If modal is opened, closing it
				if self.window:
					Utils.log('Trying to close login modal')
					self.window.close()

				# Saving username and password into the XBMC settings
				self.addon.setSetting('username', str(self.username))
				self.addon.setSetting('password', str(self.password))

				return True
			else:
				Utils.log('Login failed')
				xbmcgui.Dialog().ok(default.language(30015), '', default.language(30017), default.language(30018))

				if not self.window:
					self.window = AuthForm(default.language(30006))
					self.window.doModal()

				return False
		except:
			Utils.log('Something nasty happened during login (' + str(sys.exc_info()[0]) + ')')
			xbmc.executebuiltin('Notification(An error occurred, Something nasty happened. Sorry...,3000,' + addon_path + '/resources/images/error.png)')
			return False

	def Logout(self):
		self.addon.setSetting('password', '')
		default.Api.Logout()

		return True


resources/lib/Utils.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
   Ananoos plugin for XBMC
   Copyright (C) 2014 Axel de Vignon
   License: Apache 2.0 (See license.txt)
'''


import os
import re
import sys
import xbmc
import urllib
import urllib2
import json
import cookielib
from cookielib import CookieJar, DefaultCookiePolicy

default = sys.modules['__main__']

def getUserAgent():
	return default.plugin + ' (XBMC/' + default.xbmc_version + ', BuildDate: ' + default.xbmc_date + ')'


def encode(string):
	try:
		return string.encode('utf-8','replace')
	except:
		log('ENCODE() - Failed to encode to UTF-8')

	try:
		return string.encode('latin','replace')
	except:
		log('ENCODE() - Failed to encode to latin')

	try:
		return string.decode(ENCODING,'replace').encode('utf-8','replace')
	except:
		log('ENCODE() - Failed to encode to UTF-8 (alt-method) - using repr()')
		ret = repr(string)
		if ret.startswith('u'): ret = ret[1:]
		return ret[1:-1]


def error(message):
	log(message)
	return str(sys.exc_info()[1])

def log(message):
	# Hidding password into the logs
	message = re.sub('password=[a-zA-Z0-9_-]+', 'password=[hidden]', message)

	print default.plugin + ': %s' % encode(str(message))
	xbmc.log("%s: %s" % (default.plugin,
                         message),
                         level=xbmc.LOGDEBUG)



# Generating Auth Dialog
from resources.lib.pyxbmct.addonwindow import *
class AuthForm(AddonDialogWindow):

	def __init__(self, title=''):
		# You need to call base class' constructor.
		super(AuthForm, self).__init__(title)
		# Set the window width, height and the grid resolution: 2 rows, 3 columns.
		self.setGeometry(600, 350, 7, 3)

		self.__setControls()
		self.__setNavigation()

	def __setControls(self):
		# introduction message
		text1 = Label(default.language(30007))
		self.placeControl(text1, 0, 0, columnspan=3)
		text2 = Label(default.language(30008))
		self.placeControl(text2, 1, 0, columnspan=3)
		text3 = Label(default.language(30009))
		self.placeControl(text3, 2, 0, columnspan=3)

		# form
		label1 = Label(default.language(30012))
		self.placeControl(label1, 4, 0)
		self.username = Edit('username')
		self.placeControl(self.username, 4, 1, columnspan=2)
		label2 = Label(default.language(30002))
		self.placeControl(label2, 5, 0)
		self.password = Edit('password', isPassword=1)
		self.placeControl(self.password, 5, 1, columnspan=2)

		self.username.setText(default.addon.getSetting('username'))
		self.password.setText(default.addon.getSetting('password'))


		# Create a button.
		self.button = Button(default.language(30010))
		# Place the button on the window grid.
		self.placeControl(self.button, 6, 1)
		# Set initial focus on the button.
		self.setFocus(self.username)
		# Connect the button to a function.
		self.connect(self.button, lambda:self.doLogin())
		# Connect a key action to a function.
		self.connect(ACTION_NAV_BACK, self.close())

	def __setNavigation(self):
		"""Set up keyboard/remote navigation between controls."""
		self.username.controlUp(self.button)
		self.username.controlDown(self.password)
		self.password.controlUp(self.username)
		self.password.controlDown(self.button)
		self.button.controlUp(self.password)
		self.button.controlDown(self.username)

	def doLogin(self):
		if len(self.username.getText()) > 0 and len(self.password.getText()) > 0:
			default.Auth.Login(username = self.username.getText(), password = self.password.getText())
		else:
			return

	def doClose(self):
		self.close()




resources/lib/Api.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
   Ananoos plugin for XBMC
   Copyright (C) 2014 Axel de Vignon
   License: Apache 2.0 (See license.txt)
'''


import os
import re
import sys
import xbmc
import xbmcgui
import urllib
import urllib2
import json
import cookielib
from cookielib import CookieJar, DefaultCookiePolicy
import resources.lib.Utils as Utils
import tempfile

default = sys.modules['__main__']

class Api():

	def __init__(self):
		self.allowed_domains    = ["ananoos.com", ".ananoos.com"]
		self.api_url            = sys.modules['__main__'].API_URL
		self.cookiejar_dir      = default.addon.getSetting('cookiejar')
		self.cookies            = {}

		if not os.path.isdir(self.cookiejar_dir):
			self.cookiejar_dir = tempfile.mkdtemp()
			Utils.log('Creating a new Cookiejar in '+self.cookiejar_dir)
			default.addon.setSetting('cookiejar', self.cookiejar_dir)

		self.cookiejar		= self.cookiejar_dir+'/cookies.lwp'	
		Utils.log('Using cookiejar: '+self.cookiejar)

	def Browse(self, parent = "0"):
		return self.__Call(url='BrowseFiles', params={'parent': parent})

	def getStreamUrl(self, item_id = None):
		if item_id is None:
			return False

		return self.__Call(url='StreamFile', params={'file': item_id})

	def Login(self, username = None, password = None):
		if username is None or password is None:
			return False

		return self.__Call(url='Auth', params={'username': username, 'password': password})

	def getCookies(self):
		return self.cookies	

	def Logout(self):
		return self.__Call(url='Logout')

	def __Call(self, url, params = {}):

		#xbmc.executebuiltin('Notification('+default.language(30015)+', '+language(30016)+',3000,' + default.addon_path +'/resources/images/success.png)')

		# Matching URL
		if not re.search('http(s)?://', url):
			url = self.api_url + '/' + url

		Utils.log('Requesting URL : '+url+'?'+urllib.urlencode(params))

		# Running request
		try:
			# Setting cookies default policy
			policy = DefaultCookiePolicy(rfc2965=False, allowed_domains=self.allowed_domains)
			cj = cookielib.LWPCookieJar(policy=policy)

		 	# Starting the CookieJar handler
		 	opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cj)) 
		 	urllib2.install_opener(opener)

			# Starting the request
			req = urllib2.Request(url)
			req.add_header('User-Agent', Utils.getUserAgent())

			# Loading existing cookie(s) if applicable
			# Ignoring_discard  = True otherwise session cookies aren't sent
			if os.path.isfile(self.cookiejar):
				Utils.log('Loading existing cookies')
				cj.load(self.cookiejar, ignore_discard=True)

			# Running query
			response = urllib2.urlopen(req, urllib.urlencode(params))
			# Getting response
			content     = response.read()
			# Getting response informations
			infos    = response.info()

			# Checking the response content-type
			if infos['content-type']:
				# If response is a JSON, decoding JSON automatically
				if re.search('application/json', infos['content-type']):
					content = json.loads(content)

			# Saving private_key for later use
			for cookie in cj:
				self.cookies[cookie.name] = cookie.value

			# If unable to save the cookie jar file
			# Aborting execution, otherwise going into
			# An infinite loop
			if not os.path.isdir(self.cookiejar_dir):
		
				Utils.log('Warning: Cookiejar DIR does not exist') 
				xbmcgui.Dialog().ok(default.language(30023), '', default.language(30024))
				sys.exit()

			# Saving any cookie sent with the response
			# Ignoring_discard  = True otherwise session cookies aren't saved				
			cj.save(self.cookiejar, ignore_discard=True)    

			response.close()
			return content


		except urllib2.HTTPError, e:
			# Authentication issue, resetting the private key
			if e.code == 403:
				Utils.log('Got a 403 Forbidden, trying to login now')
				if default.Auth.Login():
					default.browse()

			#Utils.log('API Call error: ' + str(e))
			#xbmc.executebuiltin('Notification(An error occurred, Something nasty happened. Sorry...,3000,' + default.addon_path + '/resources/images/error.png)')

		except urllib2.URLError, e:
			Utils.log('Could not request the API (' + str(e) + ')')
			xbmc.executebuiltin('Notification(An error occurred, Something nasty happened. Sorry...,3000,' + default.addon_path + '/resources/images/error.png)')

		except:
			Utils.log('Something nasty happened during API Call (' + str(sys.exc_info()[0]) + ')')
			xbmc.executebuiltin('Notification(An error occurred, Something nasty happened. Sorry...,3000,' + default.addon_path + '/resources/images/error.png)')
			return False
